internal/extensions/resolve.go

```diff
@@ -0,0 +1,167 @@
+type extensionResolver struct {
+	// types is a types that just contains the protovalidate extensions.
+	types *protoregistry.Types
+
+	// legacyExtensionMap is a mapping from current protovalidate extensions to
+	// legacy protovalidate extensions, used for backwards compatibility. This
+	// map will not be modified, so it is safe to read concurrently.
+	legacyExtensionMap map[protoreflect.ExtensionType]protoreflect.ExtensionType
+}
+
+// newExtensionResolver creates a new extension resolver. This is only called at
+// init and will panic if it fails.
+func newExtensionResolver() extensionResolver {
+	resolver := extensionResolver{
+		types:              &protoregistry.Types{},
+		legacyExtensionMap: make(map[protoreflect.ExtensionType]protoreflect.ExtensionType),
+	}
+	resolver.register(validate.E_Field)
+	resolver.register(validate.E_Message)
+	resolver.register(validate.E_Oneof)
+	resolver.register(validate.E_Predefined)
+	resolver.registerLegacy(validate.E_Field)
+	resolver.registerLegacy(validate.E_Message)
+	resolver.registerLegacy(validate.E_Oneof)
+	return resolver
+}
+
+// register registers an extension into the resolver's registry. This is only
+// called at init and will panic if it fails.
+func (resolver extensionResolver) register(extension protoreflect.ExtensionType) {
+	if err := resolver.types.RegisterExtension(extension); err != nil {
+		//nolint:forbidigo // this needs to be a fatal at init
+		panic(err)
+	}
+}
+
+// registerLegacy creates and registers a legacy extension.
+func (resolver extensionResolver) registerLegacy(extension protoreflect.ExtensionType) {
+	fileDescriptor, err := protodesc.NewFile(&descriptorpb.FileDescriptorProto{
+		Name:    proto.String("buf/validate/validate_legacy.proto"),
+		Package: proto.String("buf.validate"),
+		Dependency: []string{
+			"buf/validate/validate.proto",
+			"google/protobuf/descriptor.proto",
+		},
+		Extension: []*descriptorpb.FieldDescriptorProto{
+			{
+				Name:     proto.String(string(extension.TypeDescriptor().Name()) + "_legacy"),
+				Number:   proto.Int32(int32(legacyExtensionIndex)),
+				Label:    descriptorpb.FieldDescriptorProto_LABEL_OPTIONAL.Enum(),
+				Type:     descriptorpb.FieldDescriptorProto_TYPE_MESSAGE.Enum(),
+				TypeName: proto.String(string(extension.TypeDescriptor().Message().FullName())),
+				Extendee: proto.String(string(extension.TypeDescriptor().ContainingMessage().FullName())),
+			},
+		},
+	}, protoregistry.GlobalFiles)
+	if err != nil {
+		//nolint:forbidigo // this needs to be a fatal at init
+		panic(err)
+	}
+	legacyExtension := dynamicpb.NewExtensionType(fileDescriptor.Extensions().Get(0))
+	resolver.register(legacyExtension)
+	resolver.legacyExtensionMap[extension] = legacyExtension
+}
+
+// resolve handles the majority of extension resolution logic. This will return
+// a proto.Message for the given extension if the message has the tag number of
+// the provided extension (or an equivalent legacy extension.) If there was no
```

Comment:
```suggestion
// the provided extension (or an equivalent legacy extension). If there was no
```

Reply:
Fixed in 9d5c4696e205f7a84d0d8ba131dd42d1ea2193fe